src/util/rocksdb/mod.rs

```diff
@@ -305,42 +320,15 @@ impl SliceTransform for NoopSliceTransform {
     }
 }
 
-fn prev_key(key: &[u8]) -> Result<Vec<u8>, String> {
-    if key.is_empty() {
-        return Err(String::from("Empty key has no previous key"));
-    }
-
-    let mut prev = key.to_vec();
-    let tail = prev.len() - 1;
-    if prev[tail] > 0 {
-        prev[tail] -= 1;
-    } else {
-        prev.truncate(tail);
-    }
-
-    Ok(prev)
-}
-
-pub fn roughly_cleanup_range(db: &DB, start_key: &[u8], end_key: &[u8]) -> Result<(), String> {
-    if start_key > end_key {
-        return Err(format!(
-            "[roughly_cleanup_in_range] start_key({:?}) should't larger than end_key({:?}).",
-            start_key, end_key
-        ));
-    }
-
-    if start_key == end_key {
-        return Ok(());
-    }
-
-    // Todo: use end_key directly after delete_files_in_range support [start, end) semantics.
-    let end = prev_key(end_key)?;
-    if start_key >= end.as_slice() {
-        return Ok(());
-    }
+pub fn roughly_cleanup_ranges(db: &DB, ranges: &[(Vec<u8>, Vec<u8>)]) -> Result<(), String> {
     for cf in db.cf_names() {
+        let mut delete_ranges = Vec::new();
+        for range in ranges {
+            assert!(range.0 <= range.1);
+            delete_ranges.push(Range::new(&range.0, &range.1));
+        }
         let handle = get_cf_handle(db, cf)?;
-        db.delete_file_in_range_cf(handle, start_key, &end)?;
+        db.delete_files_in_ranges_cf(handle, &delete_ranges, /* include_end */ false)?;
```

Comment:
Please check if delete_ranges empty before involve delete_files_in_ranges_cf.
BTW, we only need to calculate ranges once and reuse it for each cf.